File: services/activity.py
"""Activity & capacity — what's running locally + how much headroom is left.

Pi runs everything on one Mac, so "how much more can I run in parallel" is a
real question. This exposes:
  * live RAM / CPU (psutil)
  * which models Ollama currently has LOADED (the things using the GPU now)
  * active Pi agent sessions (Code / Crew, via a tiny registry) + downloads
  * Ollama's concurrency budget (max loaded models + parallel requests)
  * a headroom estimate ("you can add ~N more")
"""

from __future__ import annotations

import itertools
import logging
import os
import threading
import time
from typing import Dict, List, Optional

logger = logging.getLogger("activity")

_sessions: Dict[int, dict] = {}
_lock = threading.Lock()
_counter = itertools.count(1)

# Metal lets the GPU use ~75% of unified RAM for models by default.
_METAL_FRACTION = 0.75
# rough loaded footprints (GB) for the headroom hint
_SMALL_GB = 6.0     # ~7-8B quant
_MED_GB = 10.0      # ~14B quant
_BIG_GB = 55.0      # qwen3-coder-next etc.


# ───────────── session registry (active agent streams) ─────────────
# Every live operation in PiAOS should register here so the Activity panel can
# (a) SHOW it in real time and (b) STOP it. Each entry optionally carries a
# `cancel` callable (the subsystem's own stop primitive — jobs.cancel,
# agent_runs.stop, research cancel, etc.) so one Stop button covers everything.
def register(kind: str, model: Optional[str] = None, label: Optional[str] = None,
             ref: Optional[str] = None, cancel=None) -> int:
    with _lock:
        sid = next(_counter)
        _sessions[sid] = {"id": sid, "kind": kind, "model": model,
                          "label": (label or "")[:60], "started_at": int(time.time()),
                          "ref": ref, "_cancel": cancel}
        return sid


def unregister(sid: Optional[int]) -> None:
    if sid is None:
        return
    with _lock:
        _sessions.pop(sid, None)


def stop(entry_id: int) -> bool:
    """Stop a registered live op by its registry id. Calls the entry's own
    cancel primitive (set at register time). Returns True if a cancel ran."""
    with _lock:
        entry = _sessions.get(int(entry_id)) if str(entry_id).lstrip("-").isdigit() else None
        cancel = entry.get("_cancel") if entry else None
    if not cancel:
        return False
    try:
        cancel()
        return True
    except Exception as e:  # noqa: BLE001
        logger.warning("activity.stop(%s) cancel failed: %s", entry_id, e)
        return False


def _active() -> List[dict]:
    # Strip the non-serializable cancel callable; expose a `stoppable` flag + ref
    # so the frontend can render a Stop button and send the id back.
    with _lock:
        out = []
        for v in _sessions.values():
            out.append({"id": v["id"], "kind": v["kind"], "model": v.get("model"),
                        "label": v.get("label"), "started_at": v.get("started_at"),
                        "ref": v.get("ref"), "stoppable": v.get("_cancel") is not None})
        return out


# ───────────── ollama loaded models ─────────────
def _ollama_loaded() -> List[dict]:
    try:
        import httpx
        r = httpx.get("http://localhost:11434/api/ps", timeout=3)
        r.raise_for_status()
        out = []
        for m in r.json().get("models", []):
            sz = int(m.get("size", 0))
            vram = int(m.get("size_vram", 0))
            out.append({"name": m.get("name", ""), "size_gb": round(sz / 1e9, 1),
                        "gpu_gb": round(vram / 1e9, 1), "expires_at": m.get("expires_at")})
        return out
    except Exception as e:  # noqa: BLE001
        logger.debug("ollama /api/ps failed: %s", e)
        return []


def _downloads() -> List[dict]:
    """Active Local-AI installs / model pulls (best-effort)."""
    try:
        from routes.models_hub_routes import _JOBS, _JOBS_LOCK
        with _JOBS_LOCK:
            return [{"id": k, "msg": v.get("msg"), "pct": v.get("pct")}
                    for k, v in _JOBS.items() if v.get("state") == "running"]
    except Exception:  # noqa: BLE001
        return []


def _concurrency() -> dict:
    return {
        "max_loaded_models": os.environ.get("OLLAMA_MAX_LOADED_MODELS", "3 (default)"),
        "num_parallel": os.environ.get("OLLAMA_NUM_PARALLEL", "auto (1–4)"),
        "max_queue": os.environ.get("OLLAMA_MAX_QUEUE", "512 (default)"),
    }


def capacity() -> dict:
    import psutil
    vm = psutil.virtual_memory()
    total = vm.total / 1e9
    free = vm.available / 1e9
    used = vm.used / 1e9
    cpu = psutil.cpu_percent(interval=0.0)

    loaded = _ollama_loaded()
    loaded_gb = round(sum(m["size_gb"] for m in loaded), 1)
    gpu_budget = round(total * _METAL_FRACTION, 1)
    avail = max(0.0, min(free, gpu_budget - loaded_gb))

    active = _active()
    downloads = _downloads()

    # human advice
    if avail >= _BIG_GB:
        advice = f"~{avail:.0f} GB free for models — room for a big coder, or several smaller ones in parallel."
    elif avail >= _MED_GB:
        advice = f"~{avail:.0f} GB free — you can add ~{int(avail // _MED_GB)} more mid-size model(s) (≈14B)."
    elif avail >= _SMALL_GB:
        advice = f"~{avail:.0f} GB free — room for a small model (≈7B). Big models would queue/evict."
    else:
        advice = "Memory is tight — new model loads will queue or evict an idle model."

    return {
        "ram": {"total_gb": round(total, 1), "used_gb": round(used, 1),
                "free_gb": round(free, 1), "percent": round(vm.percent)},
        "cpu_percent": round(cpu),
        "gpu_budget_gb": gpu_budget,
        "loaded_models": loaded,
        "loaded_gb": loaded_gb,
        "available_for_models_gb": round(avail, 1),
        "headroom": {
            "small_7b": int(avail // _SMALL_GB),
            "mid_14b": int(avail // _MED_GB),
            "fits_big_coder": avail >= _BIG_GB,
        },
        "concurrency": _concurrency(),
        "active": active,
        "active_count": len(active),
        "downloads": downloads,
        "advice": advice,
    }

File: static/js/activity.js
// static/js/activity.js — live "what's running + how much headroom" indicator.
// Polls /api/activity (services/activity). Updates a glanceable badge on the
// sidebar button, and opens a panel with RAM/GPU, loaded models, active agent
// sessions, downloads, concurrency budget, and a headroom hint.

let _open = false, _interval = null, _started = false, _badgeMisses = 0;

const _el = (id) => document.getElementById(id);
const _esc = (s) => (s || '').replace(/[&<>"]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]));
const _api = (p, o) => fetch(p, Object.assign({ credentials: 'same-origin' }, o || {}));

function _ramColor(p) { return p >= 85 ? '#ff5758' : p >= 65 ? '#e0a000' : '#2ecc71'; }
function _ago(ts) { const s = Math.max(0, Math.floor(Date.now() / 1000) - (ts || 0)); return s < 60 ? s + 's' : Math.floor(s / 60) + 'm'; }

async function _fetch() { try { return await (await _api('/api/activity')).json(); } catch (e) { return null; } }

function _updateBadge(cap) {
  const badge = _el('activity-badge');
  if (!badge) return;
  // A single failed 5s poll shouldn't blink the badge off — hold the last
  // reading through transient hiccups, hide only after 3 misses in a row.
  if (!cap || !cap.ram) {
    if (++_badgeMisses >= 3) badge.style.display = 'none';
    return;
  }
  _badgeMisses = 0;
  const p = cap.ram.percent;
  badge.style.display = 'inline-flex';
  badge.style.alignItems = 'center';
  badge.style.gap = '3px';
  badge.innerHTML = `<span style="width:7px;height:7px;border-radius:50%;background:${_ramColor(p)};display:inline-block;"></span>`
    + (cap.active_count ? `<span style="font-size:9px;opacity:.75;">${cap.active_count}</span>` : '');
  badge.title = `RAM ${p}% · ${(cap.loaded_models || []).length} model(s) loaded · ${cap.active_count} agent session(s)`;
}

// runs for the whole app session so the sidebar badge stays live
function startMonitor() {
  if (_started) return; _started = true;
  const tick = async () => { const c = await _fetch(); _updateBadge(c); if (_open) _render(c); };
  tick();
  setInterval(tick, 5000);
}

function openPanel() {
  if (_open) return;
  _open = true;
  const modal = document.createElement('div');
  modal.className = 'modal';
  modal.id = 'activity-modal';
  modal.innerHTML = `
    <div class="modal-content" style="max-width:640px;width:94vw;max-height:88vh;display:flex;flex-direction:column;">
      <div style="display:flex;align-items:center;gap:10px;margin-bottom:6px;">
        <div style="font-weight:600;font-size:15px;flex:1;">Activity &amp; Capacity</div>
        <span style="font-size:11px;opacity:.55;">live · refreshes 5s</span>
        <button class="modal-close" id="activity-close" aria-label="Close">×</button>
      </div>
      <div style="font-size:12px;opacity:.7;margin-bottom:10px;">Everything runs on this Mac. Here's what's using it now and how much room is left.</div>
      <div id="activity-body" style="overflow:auto;display:flex;flex-direction:column;gap:14px;">
        <div style="opacity:.6;padding:20px;text-align:center;">Loading…</div>
      </div>
    </div>`;
  document.body.appendChild(modal);
  _el('activity-close').addEventListener('click', closePanel);
  modal.addEventListener('click', (e) => { if (e.target === modal) closePanel(); });
  document.addEventListener('keydown', _onEsc);
  _fetch().then(_render);
}

function _onEsc(e) { if (e.key === 'Escape' && _open) closePanel(); }

function _bar(usedGb, totalGb, gpuBudgetGb, pct) {
  const usedPct = Math.min(100, Math.round((usedGb / totalGb) * 100));
  const gpuPct = Math.min(100, Math.round((gpuBudgetGb / totalGb) * 100));
  return `<div style="position:relative;height:22px;border-radius:6px;background:var(--panel2,#262b34);border:1px solid var(--border,#333);overflow:hidden;">
      <div style="position:absolute;left:0;top:0;bottom:0;width:${usedPct}%;background:${_ramColor(pct)};opacity:.85;"></div>
      <div style="position:absolute;left:${gpuPct}%;top:0;bottom:0;width:2px;background:var(--fg);opacity:.5;" title="Metal GPU budget (~75%)"></div>
      <div style="position:absolute;inset:0;display:flex;align-items:center;justify-content:center;font-size:11px;font-weight:600;">${usedGb} / ${totalGb} GB used (${pct}%)</div>
    </div>
    <div style="font-size:10px;opacity:.5;margin-top:2px;">▏ marker = GPU budget for models (~${gpuBudgetGb} GB)</div>`;
}

function _section(title, inner) {
  return `<section style="border:1px solid var(--border,#333);border-radius:8px;padding:10px 12px;">
    <div style="font-size:12px;font-weight:600;margin-bottom:7px;">${title}</div>${inner}</section>`;
}

function _render(cap) {
  const body = _el('activity-body');
  if (!body) return;
  if (!cap || !cap.ram) { body.innerHTML = '<div style="opacity:.6;padding:20px;">Could not read capacity.</div>'; return; }

  const r = cap.ram;
  let html = _section('Memory', _bar(r.used_gb, r.total_gb, cap.gpu_budget_gb, r.percent)
    + `<div style="display:flex;gap:14px;font-size:12px;margin-top:8px;flex-wrap:wrap;">
        <span>🟢 Free: <b>${r.free_gb} GB</b></span>
        <span>📦 Models loaded: <b>${cap.loaded_gb} GB</b></span>
        <span>🧮 CPU: <b>${cap.cpu_percent}%</b></span>
        <span>🎮 For models: <b>${cap.available_for_models_gb} GB</b></span>
      </div>`);

  // headroom advice
  html += `<div class="call" style="border:1px solid #2ecc71;border-left:4px solid #2ecc71;border-radius:8px;padding:9px 12px;background:rgba(46,204,113,.07);font-size:13px;">
      <b style="color:#2ecc71;">Headroom</b> · ${_esc(cap.advice)}<br>
      <span style="font-size:11px;opacity:.7;">≈ ${cap.headroom.small_7b} more small (7B) · ${cap.headroom.mid_14b} more mid (14B) · big coder ${cap.headroom.fits_big_coder ? 'fits ✓' : 'too big now'}</span>
    </div>`;

  // loaded models
  const lm = cap.loaded_models || [];
  html += _section('Models loaded right now', lm.length
    ? lm.map(m => `<div style="display:flex;justify-content:space-between;font-size:12.5px;padding:3px 0;">
        <span>🧠 ${_esc(m.name)}</span><span style="opacity:.65;">${m.size_gb} GB${m.gpu_gb ? ' · ' + m.gpu_gb + ' GB GPU' : ''}</span></div>`).join('')
    : '<div style="font-size:12px;opacity:.55;">None loaded — models load on first use (~seconds), unload when idle.</div>');

  // active agent sessions — each with a Stop button (this panel is the universal
  // kill switch: it hits the op's real server-side cancel, so closing/✕ is not
  // needed and detached runs actually halt).
  const act = cap.active || [];
  const _stopBtn = (id) => `<button class="act-stop" data-stop="${id}" title="Stop this operation"
      style="margin-left:auto;font-size:10px;padding:1px 8px;border:1px solid var(--border);border-radius:6px;background:transparent;color:var(--red,#ff5758);cursor:pointer;flex-shrink:0;">Stop</button>`;
  html += _section(`Running now (${act.length})`, (act.length || (cap.downloads || []).length)
    ? (act.map(a => `<div style="font-size:12.5px;padding:3px 0;display:flex;align-items:center;gap:5px;">▶ <b>${_esc(a.kind)}</b> ${_esc(a.label || '')} <span style="opacity:.55;">· ${_esc(a.model || '')} · ${_ago(a.started_at)}</span>${a.stoppable ? _stopBtn(a.id) : ''}</div>`).join('')
       + (cap.downloads || []).map(d => `<div style="font-size:12.5px;padding:3px 0;">⬇ ${_esc(d.id.replace('ollama:', ''))} <span style="opacity:.55;">${d.pct != null ? d.pct + '%' : ''}</span></div>`).join(''))
    : '<div style="font-size:12px;opacity:.55;">Nothing running — start a Chat, Code, Crew, or download.</div>');

  // concurrency budget
  const c = cap.concurrency || {};
  html += _section('Parallel budget (Ollama)',
    `<div style="font-size:12.5px;line-height:1.7;">
       • Up to <b>${_esc(String(c.max_loaded_models))}</b> models loaded at once<br>
       • Up to <b>${_esc(String(c.num_parallel))}</b> parallel requests per model<br>
       • Extra requests queue (max ${_esc(String(c.max_queue))}) — they don't fail, just wait their turn.
     </div>
     <div style="font-size:11px;opacity:.6;margin-top:6px;">You can run several chats/Code/Crew sessions together; the single GPU time-slices, so each is a bit slower under heavy load.</div>`);

  body.innerHTML = html;

  // Wire each Stop button → unified /api/activity/stop, then refresh immediately
  // (the 5s poll would otherwise leave a stopped row lingering).
  body.querySelectorAll('.act-stop').forEach((btn) => {
    btn.addEventListener('click', async () => {
      const id = Number(btn.dataset.stop);
      btn.disabled = true; btn.textContent = 'Stopping…';
      try {
        await _api('/api/activity/stop', {
          method: 'POST', headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ id }),
        });
      } catch (e) {}
      try { _render(await _fetch()); } catch (e) {}
    });
  });
}

function closePanel() { if (!_open) return; _open = false; document.removeEventListener('keydown', _onEsc); const m = _el('activity-modal'); if (m) m.remove(); }
function togglePanel() { _open ? closePanel() : openPanel(); }
function isPanelOpen() { return _open; }

const activityModule = { openPanel, closePanel, togglePanel, isPanelOpen, startMonitor };
export default activityModule;
